src/l2_tracker/rewards_contract.cpp

```diff
@@ -151,3 +151,85 @@ std::vector<RewardsLogEntry> RewardsContract::Logs(uint64_t height) {
+    callData.contractAddress         = contractAddress;
+    callData.data                    = utils::getFunctionSignature("serviceNodes(uint64)") + indexABI;
+    nlohmann::json     callResult    = provider->callReadFunctionJSON(callData, blockNumber);
+    const std::string& callResultHex = callResult.get_ref<nlohmann::json::string_t&>();
+    std::string_view   callResultIt  = utils::trimPrefix(callResultHex, "0x");
+
+    const size_t        U256_HEX_SIZE                  = (256 / 8) * 2;
+    const size_t        BLS_PKEY_XY_COMPONENT_HEX_SIZE = 32 * 2;
+    const size_t        BLS_PKEY_HEX_SIZE              = BLS_PKEY_XY_COMPONENT_HEX_SIZE + BLS_PKEY_XY_COMPONENT_HEX_SIZE;
+    const size_t        ADDRESS_HEX_SIZE               = 32 * 2;
+
+    ContractServiceNode result                   = {};
+    size_t              walkIt                   = 0;
+    std::string_view    totalSize                = callResultIt.substr(walkIt, U256_HEX_SIZE);     walkIt += totalSize.size();
+    std::string_view    nextHex                  = callResultIt.substr(walkIt, U256_HEX_SIZE);     walkIt += nextHex.size();
+    std::string_view    prevHex                  = callResultIt.substr(walkIt, U256_HEX_SIZE);     walkIt += prevHex.size();
+    std::string_view    recipientHex             = callResultIt.substr(walkIt, ADDRESS_HEX_SIZE);  walkIt += recipientHex.size();
+    std::string_view    pubkeyHex                = callResultIt.substr(walkIt, BLS_PKEY_HEX_SIZE); walkIt += pubkeyHex.size();
+    std::string_view    leaveRequestTimestampHex = callResultIt.substr(walkIt, U256_HEX_SIZE);     walkIt += leaveRequestTimestampHex.size();
+    std::string_view    depositHex               = callResultIt.substr(walkIt, U256_HEX_SIZE);     walkIt += depositHex.size();
+    assert(walkIt == callResultIt.size());
+
+    // NOTE: Deserialize linked list
+    result.next                = utils::fromHexStringToUint64(nextHex);
+    result.prev                = utils::fromHexStringToUint64(prevHex);
+
+    // NOTE: Deserialise recipient
+    const size_t ETH_ADDRESS_HEX_SIZE = 20 * 2;
+    std::vector<unsigned char> recipientBytes = utils::fromHexString(recipientHex.substr(recipientHex.size() - ETH_ADDRESS_HEX_SIZE, ETH_ADDRESS_HEX_SIZE));
+    assert(recipientBytes.size() == result.recipient.max_size());
+    std::memcpy(result.recipient.data(), recipientBytes.data(), recipientBytes.size());
+
+    result.pubkey = std::string(pubkeyHex);
+
+    // NOTE: Deserialise metadata
+    result.leaveRequestTimestamp = utils::fromHexStringToUint64(leaveRequestTimestampHex);
+    result.deposit               = depositHex;
+    return result;
+}
+
+std::vector<uint64_t> RewardsContract::getNonSigners(const std::vector<std::string>& bls_public_keys) {
+    const uint64_t service_node_sentinel_id = 0;
+    ContractServiceNode service_node_end = serviceNodes(service_node_sentinel_id);
```

Comment:
Interesting point! This seems like part of a bigger problem because the non signers list needs to be in sync with the smart contracts latest data for the transaction to succeed. If the list changes while we are traversing then we will need to throw away the old data, keeping the height the same will result in an invalid tx when you go to submit.

To go even further if you dont immediately submit this transaction after getting this data, and the smart contract is updated then the non signers data becomes invalid. The BLS signature might still be correct and valid, but the non signers list must always reflect the difference between who signed and the latest smart contract list.

However, the smart contract list changes slowly, so the likelihood of it changing during this process is relatively low. Therefore, it might not be an immediate concern. If a user encounters this problem, its probably reasonable to simply ask them to run the query again.

Reply:
Yeah that's a nasty problem. We can make the function call atomic (e.g. force the EVM to return the array of all the service nodes) in one shot so atleast the data is consistent.

But we probably also need additional checks that's like just as we compute it, we check if the contract hash has changed. If it has allow the operation to be retried 2-3 times to produce a command here the contract hash has remained the same

Then output that result to the user if it's like a BLS aggregation request or error and say that the contract was modified since requesting the call and the user has to try again.

I'd agree and say it's unlikely, and it can be held off as a  _nice to have_ if we have time. At the least though when we revisit the front-end or the BLS daemon RPC commands we can log a notice to the user that there's a possibility that the command may be invalidated inbetween the time it was generated and submitted because of this possibility just as a heads up on why someone might need to repeat the step.